### PR TITLE
core(lhr): overhaul LHR details, introduce details.summary

### PR DESCRIPTION
**Pull request by paulirish**
Alright, #4384 laid the groundwork here and I've built off that for round two in the goal of _**"making the juicy details in LHR easy to consume."**_

(This is a super-breaking change. Yes.)

### Changes
* completely new shape of `details`:
  * `result.details.headings` describes the expected report columns
  * `result.details.items` contains row information, and may include _more_ data than referenced by the headings. it's the new `extendedInfo`. :)
* a new `result.details.summary` obj that includes whatever toplevel numeric summary we have. So far just `wastedMs` and `wastedKb`, but I could see `displayValue` migrating into it.

I think this is easiest to evaluate by looking at the LHR json. Here's one audit result, before and after:

🔎  Also here's a complete LHR: [tinyhouse.json.txt](https://github.com/GoogleChrome/lighthouse/files/1776823/tinyhouse.json.txt)

@patrickhulce @brendankenny  Thoughts on the new LHR shape?

cc @benschwarz @denar90 

------- 

This is a chain of PRs. Landing order is... 1st: `newdetails` (#4616). 2nd: `shallowCategories` (#4711). 3rd: `scoring2.0` (#4690)

ref #4614

### CODE REVIEW

File: lighthouse-core/report/v2/renderer/details-renderer.js [hunk 1]
```diff
@@ -315,7 +337,7 @@ class DetailsRenderer {
    */
   _renderCode(details) {
     const pre = this._dom.createElement('pre', 'lh-code');
-    pre.textContent = details.text;
+    pre.textContent = details.text || details.value; // TODO, pick a winner
```
Comment (patrickhulce) [inline lighthouse-core/report/v2/renderer/details-renderer.js]: +1 for `value`

File: lighthouse-core/report/v2/renderer/details-renderer.js [hunk 2]
```diff
@@ -165,15 +169,16 @@ class DetailsRenderer {
    * @param {!DetailsRenderer.ThumbnailDetails} value
    * @return {!Element}
    */
-  _renderThumbnail(value) {
-    if (/^image/.test(value.mimeType) === false) {
-      return this._dom.createElement('span');
-    }
+  _renderThumbnail(details) {
+    // TODO: use some empty details test instead? to handle data uris?
```
Comment (patrickhulce) [inline lighthouse-core/report/v2/renderer/details-renderer.js]: yeah let's strip out all data URIs at the audit base-class, they can make the report super beefy anyway

File: lighthouse-core/report/v2/renderer/performance-category-renderer.js [hunk 3]
```diff
@@ -45,13 +45,13 @@ class PerformanceCategoryRenderer extends CategoryRenderer {
    * @return {!Element}
    */
   _renderPerfHintAudit(audit, scale) {
-    const extendedInfo = /** @type {!PerformanceCategoryRenderer.PerfHintExtendedInfo}
-        */ (audit.result.extendedInfo);
+    const summaryInfo = /** @type {!PerformanceCategoryRenderer.PerfHintSummaryInfo}
+        */ (audit.result.summary);
```
Comment (patrickhulce) [inline lighthouse-core/report/v2/renderer/performance-category-renderer.js]: I might be missing something, I thought we were nesting `summary` inside of `details`?

File: lighthouse-core/scoring.js [hunk 4]
```diff
@@ -53,15 +53,14 @@ class ReportScoring {
           result.informative = true;
         }
 
-        return Object.assign({}, audit, {result, score: auditScore});
```
Comment (patrickhulce) [inline lighthouse-core/scoring.js]: while we're in the spirit of breaking changes, let's go whole hog and do it at the audit class level :D

**Comment by denar90** [comment]
Wow, big amount of work! Thanks for noticing 👍
Is there any v3 roadmap issue? I just found only 2 pr's but no doubt there will be more)

**Comment by paulirish** [comment]
> Is this what we currently have?

Yes, that looks mostly accurate. I'm almost done with typechecking the JSON against an iteration of these interfaces. ;)

----------

In the case of PSI audits we have an even simpler details story:

```ts
interface AuditDetails {
  summary?: AuditDetailsSummary  // could be also be a sibling to details..
  headings?: AuditTableHeading[]
  items?: AuditDetailsItem[]
}

interface AuditDetailsSummary {
  wastedMs?: number // maybe these two within a `performanceOpportunity` object?
  wastedKb?: number
}

interface AuditTableHeading {
  key: string
  text: string
  itemType: string
  granularity?: string
  displayUnit?: string
}

interface AuditDetailsItem {
  [key: string]: string | number
}
```

**Comment by paulirish** [comment]
@patrickhulce this is ready for review.

File: lighthouse-core/audits/byte-efficiency/uses-long-cache-ttl.js [hunk 5]
```diff
@@ -18,7 +18,7 @@ const IGNORE_THRESHOLD_IN_PERCENT = 0.925;
 const SCORING_POINT_OF_DIMINISHING_RETURNS = 4; // 4 KB
 const SCORING_MEDIAN = 768; // 768 KB
 
-class CacheHeaders extends ByteEfficiencyAudit {
+class CacheHeaders {
```
Comment (patrickhulce) [inline lighthouse-core/audits/byte-efficiency/uses-long-cache-ttl.js]: @paulirish let's make sure this extends audit

File: lighthouse-core/report/v2/renderer/details-renderer.js [hunk 6]
```diff
@@ -133,11 +137,16 @@ class DetailsRenderer {
    * @return {!Element}
    */
   _renderLink(details) {
+    console.assert(('text' in details), 'link details must have .url');
```
Comment (patrickhulce) [inline lighthouse-core/report/v2/renderer/details-renderer.js]: seems like this should be `'url' in details`?

File: lighthouse-core/report/v2/renderer/details-renderer.js [hunk 7]
```diff
@@ -209,26 +214,46 @@ class DetailsRenderer {
...
     const tbodyElem = this._dom.createChildOf(tableElem, 'tbody');
     for (const row of details.items) {
       const rowElem = this._dom.createChildOf(tbodyElem, 'tr');
-      for (const columnItem of row) {
-        const classes = `lh-table-column--${columnItem.type}`;
-        this._dom.createChildOf(rowElem, 'td', classes).appendChild(this.render(columnItem));
+      for (const heading of details.headings) {
+        const value = /** @type {number|string|!DetailsRenderer.DetailsJSON} */ (row[heading.key]);
+        if (typeof value === 'undefined') {
+          continue; // e.g. no lineNumber in this row item
```
Comment (patrickhulce) [inline lighthouse-core/report/v2/renderer/details-renderer.js]: don't we still want to create a td element even if it's empty?

File: lighthouse-core/runner.js [hunk 8]
```diff
@@ -93,9 +93,10 @@ class Runner {
       const resultsById = {};
       for (const audit of runResults.auditResults) resultsById[audit.name] = audit;
 
-      let report;
+      let cats;
```
Comment (patrickhulce) [inline lighthouse-core/runner.js]: cats

can we go `categories` though :)

**Comment by paulirish** [comment]
updated and squashed (to resolve silly conflicts)

**Comment by paulirish** [comment]
@patrickhulce lgty?

File: lighthouse-core/report/v2/renderer/category-renderer.js [hunk 9]
```diff
@@ -203,9 +203,10 @@ class CategoryRenderer {
     const auditsGroupedByGroup = /** @type {!Object<string,
         !Array<!ReportRenderer.AuditJSON>>} */ ({});
     manualAudits.forEach(audit => {
-      const group = auditsGroupedByGroup[audit.group] || [];
+      const groupId = audit.group || 'ungrouped';
```
Comment (brendankenny) [inline lighthouse-core/report/v2/renderer/category-renderer.js]: what is this doing?

File: lighthouse-core/report/v2/renderer/details-renderer.js [hunk 10]
```diff
@@ -70,29 +72,31 @@ class DetailsRenderer {
...
 
   /**
-   * @param {!DetailsRenderer.DetailsJSON} text
+   * @param {!DetailsRenderer.StringDetailsJSON} text
    * @return {!Element}
    */
   _renderTextURL(text) {
-    const url = text.text || '';
+    console.assert(!('text' in text), 'url details should not have .text');
+    console.assert('value' in text, 'url details must have .value');
```
Comment (brendankenny) [inline lighthouse-core/report/v2/renderer/details-renderer.js]: these should be checked by the type system?

File: lighthouse-core/report/v2/renderer/details-renderer.js [hunk 11]
```diff
@@ -133,11 +137,16 @@ class DetailsRenderer {
    * @return {!Element}
    */
   _renderLink(details) {
+    console.assert(('url' in details), 'link details must have .url');
+    console.assert(('text' in details), 'link details must have .text');
```
Comment (brendankenny) [inline lighthouse-core/report/v2/renderer/details-renderer.js]: same as above

File: lighthouse-core/report/v2/renderer/details-renderer.js [hunk 12]
```diff
@@ -425,3 +456,11 @@ DetailsRenderer.LinkDetailsJSON; // eslint-disable-line no-unused-expressions
...
  */
 DetailsRenderer.FilmstripDetails; // eslint-disable-line no-unused-expressions
+
+
+/** @typedef {{
+ *     wastedMs: (number|undefined),
+ *     wastedBytes: (number|undefined),
+ * }}
+ */
+DetailsRenderer.DetailsSummary; // eslint-disable-line no-unused-expressions
```
Comment (brendankenny) [inline lighthouse-core/report/v2/renderer/details-renderer.js]: this seems more specific than a just a "summary"?

**Comment by brendankenny** [comment]
if we want to get really fancy (and if we do TS checking of the report renderer), we can use [discriminated unions](https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions) which will type check based on the value of `type` in a particular branch. Will need to be in an externs file, though, because closure will choke on `type: "text"` in the typedef

**Comment by paulirish** [comment]
> also, it looks like we won’t be able to non-awkwardly be able to get read of text.text since so many other details types still include a text?

fwiw the current shape of details objects looks like this:



### detail types with 1 required field
type `text` with `.value`
type `code` with `.value`
type `thumbnail` w/  `.value`

type `ms`  with `.value`
type `bytes` w  `.value`
type `url`   w  `.value`


### detail types with >1 required field
type `link` with `.text` and `.url`
type `node` with `.snippet` `.selector` `.path`
type `card` with `.title` `.value` `.target` `.snippet`  

---

Additionally, the second block of 1-field types (ms/bytes/url) are currently never expressed in the LHR explicitly; they are all created via `headings..itemType`.

**Pull request merged**